runelite-client/src/main/java/net/runelite/client/plugins/grandexchange/GrandExchangePlugin.java

```diff
@@ -441,11 +517,138 @@ public void onWidgetLoaded(WidgetLoaded event)
+				String otherName = itemName.replace('-', ' ');
+				if (!itemName.contains("-") || fuzzy.fuzzyScore(itemName, input) >= fuzzy.fuzzyScore(otherName, input))
+				{
+					indices = findFuzzyIndices(itemName, input);
+				}
+				else
+				{
+					indices = findFuzzyIndices(otherName, input);
+				}
+			}
+			else
+			{
+				indices = findSubstringIndices(itemName, input);
+			}
+			Collections.reverse(indices);
+
+			StringBuilder newItemName = new StringBuilder(itemName);
+			for (int index : indices)
+			{
+				if (wasFuzzySearch && (itemName.charAt(index) == ' ' || itemName.charAt(index) == '-'))
+				{
+					continue;
+				}
+				newItemName.insert(index + 1, "</u>");
+				newItemName.insert(index, underlineTag);
+			}
+
+			itemNameWidget.setText(newItemName.toString());
+		}
+	}
+
+	@Subscribe
+	public void onGrandExchangeSearched(GrandExchangeSearched event)
+	{
+		wasFuzzySearch = false;
+
+		GrandExchangeSearchMode searchMode = config.geSearchMode();
+		final String input = client.getVar(VarClientStr.INPUT_TEXT);
+		if (searchMode == GrandExchangeSearchMode.DEFAULT || input.isEmpty())
 		{
 			return;
 		}
-		rebuildGeText();
+
+		event.consume();
+
+		int resultCount = 0;
+		if (searchMode == GrandExchangeSearchMode.FUZZY_FALLBACK)
+		{
+			List<Integer> ids = IntStream.range(0, client.getItemCount())
+					.mapToObj(itemManager::getItemComposition)
+					.filter(item -> item.isTradeable() && item.getNote() == -1
+							&& item.getName().toLowerCase().contains(input))
+					.limit(MAX_RESULT_COUNT + 1)
+					.sorted(Comparator.comparing(ItemComposition::getName))
+					.map(ItemComposition::getId)
+					.collect(Collectors.toList());
+			client.setGeSearchResultIndex(0);
+			if (ids.size() > MAX_RESULT_COUNT)
+			{
+				client.setGeSearchResultCount(-1);
+				client.setGeSearchResultIds(null);
+			}
+			else
+			{
+				resultCount = ids.size();
+				client.setGeSearchResultCount(resultCount);
+				client.setGeSearchResultIds(Shorts.toArray(ids));
+			}
+		}
+
+		if (resultCount == 0)
+		{
+			FuzzyScore fuzzy = new FuzzyScore(Locale.ENGLISH);
+
+			// We do this so that for example the items "Anti-venom ..." are still at the top
+			// when searching "anti venom"
+			ToIntFunction<ItemComposition> getScore = item ->
+			{
+				int score = fuzzy.fuzzyScore(item.getName(), input);
+				if (item.getName().contains("-"))
+				{
+					return Math.max(fuzzy.fuzzyScore(item.getName().replace('-', ' '), input), score);
+				}
+				return score;
+			};
+
+			List<Integer> ids = IntStream.range(0, client.getItemCount())
+					.mapToObj(itemManager::getItemComposition)
+					.filter(item -> item.isTradeable() && item.getNote() == -1)
+					.filter(item -> getScore.applyAsInt(item) > 0)
+					.sorted(Comparator.comparingInt(getScore).reversed()
+							.thenComparing(ItemComposition::getName))
+					.limit(MAX_RESULT_COUNT)
+					.map(ItemComposition::getId)
+					.collect(Collectors.toList());
+
+			client.setGeSearchResultIndex(0);
```

Comment:
This looks like it's common between the `if (searchMode == GrandExchangeSearchMode.FUZZY_FALLBACK)` block and this block. Can it be extracted from each of them?